All_home_works/hw1/task3.py

```diff
@@ -0,0 +1,24 @@
+"""
+Classic task, a kind of walnut for you
+Given four lists A, B, C, D of integer values,
+    compute how many tuples (i, j, k, l) there are such that A[i] + B[j] + C[k] + D[l] is zero.
+We guarantee, that all A, B, C, D have same length of N where 0 ≤ N ≤ 1000.
+"""
+from typing import List
+
+
+def check_sum_of_four(a: List[int], b: List[int], c: List[int], d: List[int]) -> int:
+    result = 0
+
+    for num_a, num_b, num_c, num_d in generate_all_variant(a, b, c, d):
+        if num_a + num_b + num_c + num_d == 0:
+            result += 1
+    return result
+
+
+def generate_all_variant(a, b, c, d):
```

Comment:
What is the algirithmical complexity of this code?

Reply:
(n^4)

Reply:
Let's think about the n^3 solution.

Reply:
I think a lot about it. It's hardly imaging for me to solve it with n^3. Or maybe we will consume a lot of memory to keep n^3 numbers in it.
Is it possible?

Reply:
The expected solution is O(n^2) complexity with O(n^2) extra memory.

But at the first I want you to find the solution with O(n^3) and O(n). They use the same approach it will be easy to go from one to another. 

You need a different data structure to solve this task.

Reply:
Ok. I tried to make it not memory-consuming. But now I got it.